utils/remote-run

```diff
@@ -39,16 +39,36 @@ class CommandRunner(object):
         return subprocess.Popen(command, **kwargs)
 
     def send(self, local_to_remote_files):
+        # Directories to make on the remote target.
+        dirs_to_make = []
+
+        # Path pairs to put.
+        local_to_remote_puts = []
+
+        for local_file, remote_file in local_to_remote_files.viewitems():
+            if os.path.isDir(local_file):
+                # Directories need to be created first and the put command
+                # has the parent directory as the remote part.
+                dirs_to_make.append(remote_file)
+                local_to_remote_puts.append((local_file,
+                                             posixpath.dirname(remote_file))
+                assert(posixpath.basename(local_file)
+                    == posixpath.basename(remote_file))
+            else:
+                # Files need their parent directory to be created and the put
+                # command can have the full path.
+                dirs_to_make.append(posixpath.dirname(remote_file))
+                local_to_remote_puts.append((local_file, remote_file))
+
         # Prepare the remote directory structure.
-        # FIXME: This could be folded into the sftp connection below.
-        dirs_to_make = self._dirnames(local_to_remote_files.viewvalues())
```

Comment:
`_dirnames` was uniquing too, which I suspect is why all the tests are now failing.